rfc/0001-policy-server-custom-resource.md

```diff
@@ -0,0 +1,426 @@
+running on the same cluster; each one with its own dedicated configuration.
+
+The second change, introducing a relation between ClusterAdmissionPolicy and
+the freshly created PolicyServer resource, allows the Kubernetes administrators
+to create Policy Server Deployments that are reserved to a subset of chosen policies.
+This makes the whole infrastructure more resilient and scalable.
+
+## Examples / User Stories
+[examples]: #examples
+
+> Examples of how the feature will be used. Interactions should show the action and the response. When appropriate, provide user stories in the form of "As a [role], I want [feature], so [that]."
+
+As a Kubernetes Operator, I want to allocate a dedicated Policy Server Deployment,
+so that "noisy" Namespaces/Tenants generating lots of policy evaluations
+are isolated from the rest of the cluster and do not affect other users.
+
+As a Kubernetes Operator, I want to allocate a dedicated Policy Server Deployment,
+so that I can run mission critical policies inside of this Policy Server "pool",
+making my whole infrastructure more resilient.
+
+As a Kubernetes Operator, I want to tune the deployment settings of Policy Server,
+by editing a clearly defined Custom Resource.
+
+
+As a Kubernetes Operator managing multiple clusters via fleet,
+I want to have a consistent setup of Kubewarden across all my clusters,
+but I want to express that in an easy way,
+so that I can simplify the process of maintaining the fleet configuration files.
+
+# Detailed design
+[design]: #detailed-design
+
+> This is the bulk of the RFC. Explain the design in enough detail for somebody familiar with the product to understand, and for somebody familiar with the internals to implement.
+> 
+> This section should cover architecture aspects and the rationale behind disruptive technical decisions (when applicable), as well as corner-cases and warnings.
+
+## Kubernetes Resources
+
+A new Custom Resource is going to be defined: `PolicyServer`. This CR will
+hold all the configuration details of the Policy Server Deployment.
+
+This is just a draft proposal of the CR:
+
+```yaml
+apiVersion: policies.kubewarden.io/v1alpha2
+kind: PolicyServer
+metadata:
+  name: reserved-instance-for-tenant-a
+spec:
+  image: ghcr.io/kubewarden/policy-server:v1.0.0
+  replicaSize: 2
+  sources:
+    insecure:
+    - insecure1.registry.foo.bar
+    - insecure2.registry.foo.bar
+    secure:
+    - name: self-signed1.registry.com
+      certificate: <base64 blob>
+    - name: self-signed2.registry.com
+      certificateFrom:
+        configMapKeyRef:
+          name: name-of-the-configmap
+          key: ca_key_name
+  env:
+  - name: KUBEWARDEN_LOG_LEVEL
+    value: debug
+  - name: KUBEWARDEN_LOG_FMT
+    value: jaeger
+  annotations:
+    sidecar.jaegertracing.io/inject: default
+```
+
+The `PolicyServer` CR will be a cluster-wide resource. Take makes each
+`PolicyServer` resource identifiable by its unique `name`.
```

Comment:
Can we make the PolicyServer CR namespace scoped? I think we will need that if we want to use owner references. Owner references link resources within the same namespace, so I'm not sure if that will work for cluster scoped resources. In the end the controller, Policy Server deployment, services and configmaps will be in a namespace, so having the CR in the same namespace might be ok.

Reply:
The `ClusterAdmissionPolicy` resource, which will belong to a `PolicyServer`, is cluster-wide. The same applies to the `ValidatingWebhookConfiguration` and the `MutatingWebhookConfiguration` ones. These are the high level resources he should bind together.

This binding should definitely be possible, because they are all cluster-wide.

As for the namespaced resources, we have to figure out if the binding can be established. If this proves to not be possible, we can still manage them. We could even leave them behind when the cluster-wide resources are removed and everything would just keep working (ofc, this is not what we want to do).

Reply:
I was thinking of making both ClusterAdmissionPolicy and PolicyServer namespace scoped, I think binding should be possible if both are in the same namespace. We could then use owner references, which I am not sure if we could use them if the resources are cluster wide

Reply:
We have plans for introducing a namespaced `AdmissionPolicy` in the future, so an administrator can grant privileges at the namespace level that users can operate on, creating/modifying/deleting policies of their own I think we should not change the layout of namespaced/cluster-scoped just to get advantage of owner references.

Reply:
ah ok, thanks for the clarification